### PR TITLE
Apply peformance tuning to new sandboxes also

### PR DESCRIPTION
**Pull request by evol262**
relates to https://github.com/moby/moby/issues/35082, https://github.com/moby/libnetwork/pull/2491

Previously, values for expire_quiescent_template, conn_reuse_mode,
and expire_nodest_conn were set only system-wide. Also apply them
for new lb_* and ingress_sbox sandboxes, so they are appropriately
propagated

Signed-off-by: Ryan Barry <rbarry@mirantis.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/controller.go [hunk 1]
```diff
@@ -52,23 +52,23 @@ import (
...
+	"github.com/docker/libnetwork/diagnostic"
+	"github.com/docker/libnetwork/discoverapi"
+	"github.com/docker/libnetwork/driverapi"
+	"github.com/docker/libnetwork/drvregistry"
+	"github.com/docker/libnetwork/hostdiscovery"
+	"github.com/docker/libnetwork/ipamapi"
+	"github.com/docker/libnetwork/netlabel"
+	"github.com/docker/libnetwork/options"
+	"github.com/docker/libnetwork/osl"
+	"github.com/docker/libnetwork/types"
```
Comment (thaJeztah) [inline libnetwork/controller.go:71]: These should not be updated; the master branch uses libnetwork from this repository (only the 20.10 branch still uses the code from github.com/docker/libnetwork)

File: libnetwork/controller.go [hunk 2]
```diff
@@ -689,7 +716,7 @@ func (c *controller) NewNetwork(networkType, name string, id string, options ...
 
 	if id != "" {
 		c.networkLocker.Lock(id)
-		defer c.networkLocker.Unlock(id) // nolint:errcheck
+		defer c.networkLocker.Unlock(id)
```
Comment (thaJeztah) [inline libnetwork/controller.go:719]: Looks like an unrelated change; can you add this back?

File: libnetwork/controller.go [hunk 3]
```diff
@@ -800,7 +827,7 @@ func (c *controller) NewNetwork(networkType, name string, id string, options ...
 
 	err = c.addNetwork(network)
 	if err != nil {
-		if _, ok := err.(types.MaskableError); ok { // nolint:gosimple
+		if _, ok := err.(types.MaskableError); ok {
```
Comment (thaJeztah) [inline libnetwork/controller.go:830]: Same here

File: libnetwork/controller.go [hunk 4]
```diff
@@ -1147,7 +1174,14 @@ func (c *controller) NewSandbox(containerID string, options ...SandboxOption) (S
...
-		// Apply operating specific knobs on the load balancer sandbox
+		err := sb.osSbox.InvokeFunc(func() {
+			sb.osSbox.ApplyOSTweaks(sb.oslTypes)
+		})
+
+		if err != nil {
+			logrus.Errorf("Failed to apply performance tuning sysctls the sandbox: %v", err)
+		}
+		// Keep this just so performance is not changed
 		sb.osSbox.ApplyOSTweaks(sb.oslTypes)
```
Comment (thaJeztah) [inline libnetwork/controller.go:1185]: I'm a bit confused why `sb.osSbox.ApplyOSTweaks(sb.oslTypes)` is called twice (using different mechanisms); could explain?

File: libnetwork/osl/namespace_linux.go [hunk 5]
```diff
@@ -171,7 +181,7 @@ func GenerateKey(containerID string) string {
 			indexStr string
 			tmpkey   string
 		)
-		dir, err := os.ReadDir(basePath())
+		dir, err := ioutil.ReadDir(basePath())
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:184]: Looks like this is reverting a recent change (bad rebase perhaps?) the master branch uses `os` (the 20.10 branches uses the old `ioutil`)

File: libnetwork/osl/namespace_linux.go [hunk 6]
```diff
@@ -338,9 +348,7 @@ func createNetworkNamespace(path string, osCreate bool) error {
 
 func unmountNamespaceFile(path string) {
 	if _, err := os.Stat(path); err == nil {
-		if err := syscall.Unmount(path, syscall.MNT_DETACH); err != nil && !errors.Is(err, unix.EINVAL) {
-			logrus.WithError(err).Error("Error unmounting namespace file")
-		}
+		syscall.Unmount(path, syscall.MNT_DETACH)
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:351]: Unrelated change, or was there a reason for ignoring the error here?

File: libnetwork/osl/namespace_linux.go [hunk 7]
```diff
@@ -407,12 +415,12 @@ func (n *networkNamespace) DisableARPForVIP(srcName string) (Err error) {
 
 	err := n.InvokeFunc(func() {
 		path := filepath.Join("/proc/sys/net/ipv4/conf", dstName, "arp_ignore")
-		if err := os.WriteFile(path, []byte{'1', '\n'}, 0644); err != nil {
+		if err := ioutil.WriteFile(path, []byte{'1', '\n'}, 0644); err != nil {
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:418]: Same here for the `os -> ioutil` revert here

File: libnetwork/osl/namespace_linux.go [hunk 8]
```diff
@@ -692,5 +700,8 @@ func (n *networkNamespace) ApplyOSTweaks(types []SandboxType) {
 		case SandboxTypeLoadBalancer:
 			kernel.ApplyOSTweaks(loadBalancerConfig)
 		}
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:702]: Looks like this curly brace shouldn't be there (both should be within the switch?)

File: libnetwork/osl/namespace_linux.go [hunk 9]
```diff
@@ -692,5 +700,8 @@ func (n *networkNamespace) ApplyOSTweaks(types []SandboxType) {
...
+		case SandboxTypeIngress:
+			kernel.ApplyOSTweaks(ingressConfig)
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:704]: It looks like both Ingress and Loadbalancer are applying the same tweaks. The `loadBalancerConfig` was already there, but looking at that variable, I see it's _only_ used in this function, but defined as a package level variable. I'm not a huge fan of package level variables (if not needed), especially if they're defined way too far away from where they're used.

To prevent the risk that they diverge (assuming they should always get the same tweaks - at least currently), perhaps either inline the settings we apply. The `CheckFn: nil` also seems a bit redundant (I think that'd be the default), so I assume we can remove those as well.
Finally, it looks like the description for `ApplyOSTweaks` no longer matches what it does (it should now mention the ingress network as well?);

```go
// ApplyOSTweaks applies linux configs on sandboxes and ingress networks.
func (n *networkNamespace) ApplyOSTweaks(types []SandboxType) {
	for _, t := range types {
		switch t {
		case SandboxTypeLoadBalancer, SandboxTypeIngress:
			kernel.ApplyOSTweaks(map[string]*kernel.OSValue{
				// disables any special handling on port reuse of existing IPVS connection table entries
				// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L25:1
				"net.ipv4.vs.conn_reuse_mode": {Value: "0"},
				// expires connection from the IPVS connection table when the backend is not available
				// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L126:1
				"net.ipv4.vs.expire_nodest_conn": {Value: "1"},
				// expires persistent connections to destination servers with weights set to 0
				// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L144:1
				"net.ipv4.vs.expire_quiescent_template": {Value: "1"},
			})
		}
	}
}
```


Or if there's advantages to defining the variable once (not sure if there is), it'd be better to move the variable closer to this function (but I'm slightly in favour of the approach above instead);

```go
var tweaks = map[string]*kernel.OSValue{
	// disables any special handling on port reuse of existing IPVS connection table entries
	// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L25:1
	"net.ipv4.vs.conn_reuse_mode": {Value: "0"},
	// expires connection from the IPVS connection table when the backend is not available
	// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L126:1
	"net.ipv4.vs.expire_nodest_conn": {Value: "1"},
	// expires persistent connections to destination servers with weights set to 0
	// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L144:1
	"net.ipv4.vs.expire_quiescent_template": {Value: "1"},
}

// ApplyOSTweaks applies linux configs on sandboxes and ingress networks.
func (n *networkNamespace) ApplyOSTweaks(types []SandboxType) {
	for _, t := range types {
		switch t {
		case SandboxTypeLoadBalancer, SandboxTypeIngress:
			kernel.ApplyOSTweaks(tweaks)
		}
	}
}
```

Comment (thaJeztah) [reply libnetwork/osl/namespace_linux.go:704]: Do you know if these are settings that can be controlled through a `com.docker.network` option on `docker network create` (https://docs.docker.com/engine/reference/commandline/network_create/#bridge-driver-options)?

Double checking if this is something that a user could have customised when creating a custom ingress network (and which we would then override with fixed defaults)

**Review by thaJeztah — CHANGES_REQUESTED** [review]

**Comment by evol262** [comment]
Whew, github didn't send me any notifications for this. Most of this look like a new neovim setup behaving badly. I'll get a cleaned up version posted shortly

File: libnetwork/osl/namespace_linux.go (hunk 6 above)
Comment (evol262) [reply libnetwork/osl/namespace_linux.go:351]: Unrelated. Likely to be a bad rebase, since this was originally written against the "old" libnetwork repo before I realized it was archived and moved this over

File: libnetwork/controller.go (hunk 4 above)
Comment (evol262) [reply libnetwork/controller.go:1185]: This is more or less due to longstanding discussions around inheritance of namespaced vs non-namespaced sysctls: https://lore.kernel.org/lkml/20160221071102.9686.63148.stgit@buzz/

Values set in outside of the namespace do not _always_ propagate, but setting it both in the ns (through `InvokeFunc->nsInvoke->netns.Set`) provides a "safety valve" against the ongoing changes in kernel namespacing.

File: libnetwork/osl/namespace_linux.go (hunk 9 above)
Comment (evol262) [reply libnetwork/osl/namespace_linux.go:704]: They can't be set through `com.docker.network`, unfortunately, but I'm in total agreement about package-level variables which only get moved once. I'll inline them.

File: libnetwork/osl/namespace_linux.go [hunk 10]
```diff
@@ -689,8 +677,18 @@ func setIPv6(path, iface string, enable bool) error {
 func (n *networkNamespace) ApplyOSTweaks(types []SandboxType) {
 	for _, t := range types {
 		switch t {
-		case SandboxTypeLoadBalancer:
-			kernel.ApplyOSTweaks(loadBalancerConfig)
+		case SandboxTypeLoadBalancer, SandboxTypeIngress:
+			kernel.ApplyOSTweaks(map[string]*kernel.OSValue{
+				// disables any special handling on port reuse of existing IPVS connection table entries
+				// more info: https://github.com/torvalds/linux/blob/master/Documentation/networking/ipvs-sysctl.txt#L25:1
```
Comment (akerouanton) [inline libnetwork/osl/namespace_linux.go:683]: Just a side note: these links are broken, ipvs-sysctl.txt as been converted into rst :wink:

Comment (thaJeztah) [reply libnetwork/osl/namespace_linux.go:683]: Ah, thanks; yes, perhaps better to use a permalink to a specific tag, e.g. https://github.com/torvalds/linux/blob/v5.15/Documentation/networking/ipvs-sysctl.rst#procsysnetipv4vs-variables

(or https://github.com/torvalds/linux/blame/v5.15/Documentation/networking/ipvs-sysctl.rst#L32 to link to the specific option)

File: libnetwork/controller.go [hunk 11]
```diff
@@ -52,23 +52,6 @@ import (
...
-	"github.com/docker/docker/libnetwork/hostdiscovery"
-	"github.com/docker/docker/libnetwork/ipamapi"
-	"github.com/docker/docker/libnetwork/netlabel"
-	"github.com/docker/docker/libnetwork/options"
-	"github.com/docker/docker/libnetwork/osl"
-	"github.com/docker/docker/libnetwork/types"
-	"github.com/docker/docker/pkg/discovery"
-	"github.com/docker/docker/pkg/plugingetter"
-	"github.com/docker/docker/pkg/plugins"
-	"github.com/docker/docker/pkg/stringid"
```
Comment (thaJeztah) [inline libnetwork/controller.go:71]: These shouldn't be removed here (bad rebase?)

File: libnetwork/osl/namespace_linux.go [hunk 12]
```diff
@@ -14,14 +14,13 @@ import (
 	"syscall"
 	"time"
 
-	"github.com/docker/docker/libnetwork/ns"
-	"github.com/docker/docker/libnetwork/osl/kernel"
-	"github.com/docker/docker/libnetwork/types"
 	"github.com/docker/docker/pkg/reexec"
+	"github.com/docker/libnetwork/ns"
+	"github.com/docker/libnetwork/osl/kernel"
+	"github.com/docker/libnetwork/types"
```
Comment (thaJeztah) [inline libnetwork/osl/namespace_linux.go:20]: These shouldn't be changed here 😅 (looks like to doesn't compile currently)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
@evol262 (sorry, thought I commented already, but seems I forgot); looks like this doesn't compile currently; can you check?
